src/IRMatch.h

```diff
@@ -45,6 +47,1366 @@ bool expr_match(Expr pattern, Expr expr, std::map<std::string, Expr> &result);
+        return Op::make(to_expr(a, state));
+    }
+};
+
+template<typename A>
+HALIDE_ALWAYS_INLINE
+UnaryOp<Not, A> operator!(A &&a) {
+    return UnaryOp<Not, A>{std::forward<A>(a)};
+}
+
+template<typename A>
+inline std::ostream &operator<<(std::ostream &s, const UnaryOp<Not, A> &op) {
+    s << "!(" << op.a << ")";
+    return s;
+}
+
+template<typename C, typename T, typename F>
+struct SelectOp {
+    struct pattern_tag {};
+    C c;
+    T t;
+    F f;
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const {
+        const Select &op = (const Select &)e;
+        return (e.node_type == Select::_node_type &&
+                c.match(*op.condition.get(), state) &&
+                t.match(*op.true_value.get(), state) &&
+                f.match(*op.false_value.get(), state));
+    }
+    template<typename C2, typename T2, typename F2>
+    HALIDE_ALWAYS_INLINE bool match(const SelectOp<C2, T2, F2> &instance, MatcherState &state) const {
+        return (c.match(instance.c, state) &&
+                t.match(instance.t, state) &&
+                f.match(instance.f, state));
+    }
+    template<typename Pattern,
+             typename = typename enable_if_pattern<Pattern>::type>
+    HALIDE_ALWAYS_INLINE bool match(const Pattern &p, MatcherState &state) const {
+        return false;
+    }
+    HALIDE_ALWAYS_INLINE Expr make(MatcherState &state) const {
+        return Select::make(to_expr(c, state), to_expr(t, state), to_expr(f, state));
+    }
+};
+
+template<typename C, typename T, typename F>
+std::ostream &operator<<(std::ostream &s, const SelectOp<C, T, F> &op) {
+    s << "select(" << op.c << ", " << op.t << ", " << op.f << ")";
+    return s;
+}
+
+template<typename C, typename T, typename F>
+HALIDE_ALWAYS_INLINE
+SelectOp<C, T, F> select(C &&c, T &&t, F &&f) {
+    return SelectOp<C, T, F>{std::forward<C>(c), std::forward<T>(t), std::forward<F>(f)};
+}
+
+template<typename A>
+struct BroadcastOp {
+    struct pattern_tag {};
+    A a;
+    int lanes;
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const {
+        if (e.node_type == Broadcast::_node_type) {
+            const Broadcast &op = (const Broadcast &)e;
+            if ((lanes == -1 || lanes == op.lanes) &&
+                a.match(*op.value.get(), state)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    template<typename A2>
+    HALIDE_ALWAYS_INLINE bool match(const BroadcastOp<A2> &op, MatcherState &state) const {
+        return a.match(op.a, state) && (lanes == op.lanes || lanes == -1 || op.lanes == -1);
+    }
+
+    HALIDE_ALWAYS_INLINE Expr make(MatcherState &state) const {
+        return Broadcast::make(to_expr(a, state), lanes);
+    }
+
+};
+
+template<typename A>
+inline std::ostream &operator<<(std::ostream &s, const BroadcastOp<A> &op) {
+    s << "broadcast(" << op.a << ")";
+    return s;
+}
+
+template<typename A>
+HALIDE_ALWAYS_INLINE
```

Comment:
An awful lot of things are marked as HALIDE_ALWAYS_INLINE rather than plain inline -- do we know if that's actually necessary?